### PR TITLE
Refresh the refernce of the closed PR when reopening

### LINKED ISSUE #24213: Closed PRs getting picked up by build servers monitoring Gitea repos with Gitea 1.19
### Description

We recently upgraded from Gitea 1.18.5 to Gitea 1.19.1, and our TeamCity server has started queuing builds for a number of closed PRs for no discernable reason. No changes are being pushed to these PRs, in fact some of the PRs's branches have been deleted.

I haven't been able to track down the specific details of what's going on, but I can see on the Gitea host filesystem that the directory under repositories/org/repo/refs/pull for the closed PR is getting re-created with the "head" file containing a commit hash that doesn't belong to that branch at all (it seems to always belong to our default 'main' branch instead).

Our only fix has been to constantly monitor TeamCity for queueing builds on closed PRs and canceling them, then fully Deleting the closed PR in Gitea. This is causing a significant problem for us at the moment, so if there's a commit I can cherry-pick to manually build and deploy a fix, we'd be very happy to do that.

This only applies to PRs that are closed, it is not happening with open or merged ones.

I found some issues/PRs that sound possibly related:
* #23707 
* #23189 

### Gitea Version

1.19.1

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Hand-built 1.19.1 with one local modification, running on an arm64 AWS EC2 instance.

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by sillyguodong**
Close #24213 
Replace #23830

#### Cause

- Before, in order to making PR can get latest commit after reopening, the `ref`(${REPO_PATH}/refs/pull/${PR_INDEX}/head) of evrey closed PR will be updated when pushing commits to the `head branch` of the closed PR.

#### Changes

- For closed PR , won't perform these behavior: insert`comment`, push `notification` (UI and email), exectue [pushToBaseRepo](https://github.com/go-gitea/gitea/blob/74225033413dc0f2b308bbe069f6d185b551e364/services/pull/pull.go#L409) function and trigger `action` any more when pushing to the `head branch` of the closed PR.
- Refresh the reference of the PR when reopening the closed PR (**even if the head branch has been deleted before**). Make the reference of PR consistent with the `head branch`.

### CODE REVIEW

File: routers/web/repo/issue.go [hunk 1]
```diff
@@ -2775,6 +2776,55 @@ func NewComment(ctx *context.Context) {
 					issue.PullRequest.HeadCommitID = ""
 					pull_service.AddToTaskQueue(issue.PullRequest)
 				}
+
+				// check whether the ref of PR <refs/pulls/pr_index> in base repo is consistent with the head commit of head branch in the head repo
```
Comment (lunny) [inline routers/web/repo/issue.go]: Why these logics are not in the reopen function but in NewComment ?

File: routers/web/repo/issue.go [hunk 2]
```diff
@@ -2765,8 +2766,8 @@ func NewComment(ctx *context.Context) {
 				pr, err = issues_model.GetUnmergedPullRequest(ctx, pull.HeadRepoID, pull.BaseRepoID, pull.HeadBranch, pull.BaseBranch, pull.Flow)
 				if err != nil {
 					if !issues_model.IsErrPullRequestNotExist(err) {
-						ctx.ServerError("GetUnmergedPullRequest", err)
-						return
+						ctx.Flash.Error(ctx.Tr("repo.issues.dependency.pr_close_blocked"))
+						ctx.Redirect(fmt.Sprintf("%s/pulls/%d", ctx.Repo.RepoLink, pull.Index))
```
Comment (lunny) [inline routers/web/repo/issue.go:2789]: need `return`

Comment (wxiaoguang) [reply routers/web/repo/issue.go:2789]: There are so many similar mistakes. What do you think about my proposal https://github.com/go-gitea/gitea/issues/24120,  the Proposal 2 ?

File: routers/web/repo/issue.go (hunk 1 above)
Comment (sillyguodong) [reply routers/web/repo/issue.go]: The buttons of `close` and `reopen` are in this form.
https://github.com/go-gitea/gitea/blob/b9a97ccd0ea1ee44db85b0fbb80b75255af7c742/templates/repo/issue/view_content.tmpl#L108


https://github.com/go-gitea/gitea/blob/b9a97ccd0ea1ee44db85b0fbb80b75255af7c742/routers/web/web.go#L1107

**Comment by silverwind** [comment]
@sillyguodong please make sure to resolve resolved conversations.

**Review by wolfogre — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Pull request merged**